oxenss/snode/service_node.cpp

```diff
@@ -1196,4 +1732,279 @@ void ServiceNode::process_push_batch(std::string_view blob, std::string_view sen
     log::trace(logcat, "Saving all: end");
 }
 
+void ServiceNode::retryable_requests_thread_entry_point() {
+    // The min and max amount of time this node will backoff between failed retry requests
+    constexpr auto MIN_RETRY_DELAY = 1s;
+    constexpr auto MAX_RETRY_DELAY = 60s;
+    constexpr auto RETRY_BACKOFF_COEFF = 1.75f;
+
+    while (!shutting_down_) {
+        // At longest, we timeout on the blocking sleep every 5s, or, as soon as someone wakes up
+        // the thread by notifying the condition var
+        //  - when a new retryable request is added
+        //  - we're shutting down
+        //  - or we know there's an earlier deadline in the list of requests to be retried
+        //  - a node's contact detail was updated
+        //  - a retryable request failed and a new deadline was posted
+        auto earliest_deadline = std::chrono::steady_clock::now() + 5s;
+
+        std::unique_lock lock{retryable_requests_mutex};
+        retryable_requests_cv.wait_until(lock, earliest_deadline);
+
+        if (shutting_down_)
+            continue;
+
+        // Log the current retries
+        auto now = std::chrono::steady_clock::now();
+        if (log::Level level = log::Level::debug;
+            log::get_level(logcat) <= level && retryable_requests.size()) {
```

Comment:
This is another invariant `level` that obfuscates the eventual log statement (by moving the invariant log level 60 lines away from the log statement).

It also is a bit weird in that the `level` variable here is tracking the invariant rather than the value, and would seem to be much more usefully written as:
```suggestion
        if (log::Level level = log::get_level(logcat); level <= log::Level::debug && retryable_requests.size()) {
```
with changes in the following code to use `level` instead of `log::get_level(logcat)`.

Note also that every call to `get_level` involves an atomic load, and so that change vastly reduces the amount of atomic loads in the code below.